michaelcheah/seldon-core
# Integration test request logger

## Overview
This PR allows one to test the behaviour of the seldon-request-logger through integration tests. The goal is that these tests can be re-used in future if the request logger is written, regardless of its implementation (as long as requests and the format of the elasticsearch inserted docs remain the same, only the docker image for the request logger needs to be changed)

## Prerequisites
1. An active seldon deploy cluster with model metadata enabled and a `seldon` namespace
2. docker and pytest docker installed locally (along with all the requirements in `components/seldon-request-logger/tests/requirements.txt`)
3. Configure a file `components/seldon-request-logger/tests/.env`. This should contain the values of your seldon deploy cluster. A template is provided at `components/seldon-request-logger/tests/.env.template`

## How to run
This assumes all commands are run from `components/seldon-request-logger`
1. `pip install -r tests.requirements.txt`
2. `pytest -vvvx` - this runs all tests with maximum verbosity, but will exit immediately if any test fails. Use `pytest -vvvsx` for test debug prints

### How it works
For all test sessions, 
* Docker compose file is run by `pytest-docker` which starts a new instance of both the `seldon-request-logger` and `elasticsearch`. These are local, and are exposed via localhost so have nothing to do with any deployed cluster and can be safely brought up and down between tests.
* All tests follow a similar structure where a "scenario" is tested. These scenarios each have 
  1. The input request content
  2. Input request headers
  3. Expected elasticsearch index
  4. Expected elasticsearch docs
* For all test scenarios, we check the following:
  1. Can the input request be posted successfully?
  2. Is the expected elasticsearch index created for the given input request
  3. Can every doc in the expected elasticsearch index be retrieved?
  4. For every doc in this index, do their fields match the expected elasticsearch docs in our scenario
* For tests with metadata, there is the added complication of requiring both the `predictionSchema` and a running deployment using the model to be present on the seldon deploy cluster. Thus for those tests, we 
  1. Create model metadata for the models we want to test (we test Iris, CIFAR10, movie sentiment, and income datasets here)
  2. Create a test deployment for each of these models, and wait for them to be running before sending the scenario's input request.
  3. Proceed with the usual checks

## Troubleshooting
* If a test is killed prematurely, this might leave some docker containers still running. Check this with `docker ps -a`. This prevents the tests from being re-run. Kill and remove these containers as necessary
* elasticsearch docker has a known issue with virtual memory that can be solved https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html
